compiler/hash-ast/src/operator.rs

```diff
@@ -1,37 +1,201 @@
-//! Hash compiler AST operator abstract representations.    
+//! Hash compiler AST operator abstract representations.
+//!
+//! All rights reserved 2022 (c) The Hash Language authors
 
-use std::borrow::Cow;
+/// Struct representing an operator with a kind and a flag
+/// denoting whether the operator is re-assigning the left
+/// hand side expression.
+#[derive(Debug, Clone)]
+pub struct Operator {
+    /// The kind of operator.
+    pub kind: OperatorKind,
+    /// Flag representing where the operator is re-assigning or not.
+    pub assignable: bool,
+}
+
+#[derive(Clone, Copy)]
+/// The operator kind enumeration. This represents all types of
+/// operator present in the language.
+pub enum OperatorKind {
+    /// '=='
+    EqEq,
+    /// '!='
+    NotEq,
+    /// '|'
+    BitOr,
+    /// '||'
+    Or,
+    /// '&'
+    BitAnd,
+    /// '&&'
+    And,
+    /// '^'
+    BitXor,
+    /// '^^'
+    Exp,
+    /// '>'
+    Gt,
+    /// '>='
+    GtEq,
+    /// '<'
+    Lt,
+    /// '<='
+    LtEq,
+    /// '>>'
+    Shr,
+    /// '<<'
+    Shl,
+    /// '+'
+    Add,
+    /// '-'
+    Sub,
+    /// '*'
+    Mul,
+    /// '/'
+    Div,
+    /// '%'
+    Mod,
 
-// Types of unary operators that might need to be transformed.
-pub enum UnaryOpType {
-    FnCall(&'static str),
-    Ref,
-    Deref,
+    /// 'as'
+    // @@Cleanup: This is a special operator since it performs a higher order operation compared
+    //            to all other members of the Operator enum, maybe it should be in it's own enum??
```

Comment:
Good point, what does "higher order" mean. You mean like it is not just between expressions?

Reply:
Yeah, it's kind of strange because it's like on an operator but it receives special handling when parsing. Take the code snippet below:
https://github.com/hash-org/lang/blob/79961772ac1854b626d0650a981474dcbdb3a664/compiler/hash-parser/src/gen.rs#L2660-L2673

All other operators go through a particular transformation where they're transpiled, whereas for the `as` operator, the expression gets converted into an `ExpressionKind::Typed`.  This is fine, however, it could somehow be improved to not contain such special handling, especially if we add operators of a similar nature.

Reply:
Fair enough, I guess it is because it always has the lowest precedence and also takes a type and an expression? Maybe that's what should be stated.

Reply:
Yeah